2004-11-18  Andrew Cagney  <cagney@gnu.org>

        * dwarf2read.c: Include "top.c".
        (dwarf2_has_info): Check for readnever_symbol_files.
        * symfile.c (readnever_symbol_files): Define.
        * top.h (readnever_symbol_files): Declare.
        * main.c (captured_main): Add --readnever option.
        (print_gdb_help): Ditto.

2004-11-18  Andrew Cagney  <cagney@gnu.org>

        * gdb.texinfo (File Options): Document --readnever.

Index: gdb-6.8.50.20090811/gdb/doc/gdb.texinfo
===================================================================
--- gdb-6.8.50.20090811.orig/gdb/doc/gdb.texinfo	2009-08-13 10:14:15.000000000 +0200
+++ gdb-6.8.50.20090811/gdb/doc/gdb.texinfo	2009-08-13 10:14:18.000000000 +0200
@@ -991,6 +991,12 @@ Read each symbol file's entire symbol ta
 the default, which is to read it incrementally as it is needed.
 This makes startup slower, but makes future operations faster.
 
+@item --readnever
+@cindex @code{--readnever}
+Do not read each symbol file's symbolic debug information.  This makes
+startup faster but at the expense of not being able to perform
+symbolic debugging.
+
 @end table
 
 @node Mode Options
Index: gdb-6.8.50.20090811/gdb/main.c
===================================================================
--- gdb-6.8.50.20090811.orig/gdb/main.c	2009-08-13 10:14:15.000000000 +0200
+++ gdb-6.8.50.20090811/gdb/main.c	2009-08-13 10:14:18.000000000 +0200
@@ -382,6 +382,7 @@ captured_main (void *data)
       {"xdb", no_argument, &xdb_commands, 1},
       {"dbx", no_argument, &dbx_commands, 1},
       {"readnow", no_argument, &readnow_symbol_files, 1},
+      {"readnever", no_argument, &readnever_symbol_files, 1},
       {"r", no_argument, &readnow_symbol_files, 1},
       {"quiet", no_argument, &quiet, 1},
       {"q", no_argument, &quiet, 1},
@@ -1030,6 +1031,7 @@ Options:\n\n\
   fputs_unfiltered (_("\
   --quiet            Do not print version number on startup.\n\
   --readnow          Fully read symbol files on first access.\n\
+  --readnever        Do not read symbol files.\n\
 "), stream);
   fputs_unfiltered (_("\
   --se=FILE          Use FILE as symbol file and executable file.\n\
Index: gdb-6.8.50.20090811/gdb/symfile.c
===================================================================
--- gdb-6.8.50.20090811.orig/gdb/symfile.c	2009-08-13 10:14:15.000000000 +0200
+++ gdb-6.8.50.20090811/gdb/symfile.c	2009-08-13 10:14:18.000000000 +0200
@@ -79,6 +79,7 @@ static void clear_symtab_users_cleanup (
 
 /* Global variables owned by this file */
 int readnow_symbol_files;	/* Read full symbols immediately */
+int readnever_symbol_files;	/* Never read full symbols.  */
 
 /* External variables and functions referenced. */
 
Index: gdb-6.8.50.20090811/gdb/dwarf2read.c
===================================================================
--- gdb-6.8.50.20090811.orig/gdb/dwarf2read.c	2009-08-13 10:14:15.000000000 +0200
+++ gdb-6.8.50.20090811/gdb/dwarf2read.c	2009-08-13 10:14:29.000000000 +0200
@@ -53,6 +53,7 @@
 #include "typeprint.h"
 #include "vec.h"
 #include "block.h"
+#include "top.h"
 
 #include <fcntl.h>
 #include "gdb_string.h"
@@ -1245,7 +1246,8 @@ dwarf2_has_info (struct objfile *objfile
 
       bfd_map_over_sections (objfile->obfd, dwarf2_locate_sections, NULL);
     }
-  return (dwarf2_per_objfile->info.asection != NULL
+  return !readnever_symbol_files
+      && (dwarf2_per_objfile->info.asection != NULL
 	  && dwarf2_per_objfile->abbrev.asection != NULL);
 }
 
Index: gdb-6.8.50.20090811/gdb/top.h
===================================================================
--- gdb-6.8.50.20090811.orig/gdb/top.h	2009-01-03 06:57:53.000000000 +0100
+++ gdb-6.8.50.20090811/gdb/top.h	2009-08-13 10:14:18.000000000 +0200
@@ -59,6 +59,7 @@ extern void set_prompt (char *);
 
 /* From random places.  */
 extern int readnow_symbol_files;
+extern int readnever_symbol_files;
 
 /* Perform _initialize initialization */
 extern void gdb_init (char *);
